clients/src/main/java/org/apache/kafka/common/protocol/Errors.java

```diff
@@ -495,7 +496,14 @@ public ApiException build(String message) {
             public ApiException build(String message) {
                 return new OperationNotAttemptedException(message);
             }
-        });
+    }),
+    KAFKA_STORAGE_ERROR(56, "Disk error when trying to access log file on the disk.",
```

Comment:
I guess I didn't understand Ismael's suggestion w.r.t. the protocol version..

In my opinion, the long term approach is to have a new retriable exception, named as either `UnknownException` or `RetriableUnknownException` such that an known error code will be translated to this exception on the client side.

One short term solution we can do in this patch is to transform KafkaStorageException to the error code of NotLeaderForPartitionException in ProduceResponse and FetchResopnse. This is hacky. But it should address this problem without causing any additional concern since producer/consumer shouldn't really care whether this is KafkaStorageException as long as it is retriable.